### PR TITLE
Lab03

### PR DESCRIPTION
**Pull request by arthurdouillard**
- update labs 03 recsys to tensorflow 2.0

### CODE REVIEW

**Comment by charlesollion** [comment]
@ogrisel you might want to review this one!

**Comment by arthurdouillard** [comment]
@ogrisel I tried, as suggested, to add a bias (embeddings layer of latent dim of 1) for both user and item after the cosine similarity, but it barely improves the test auc (+0.3pts) thus it wasn't added finally.

File: labs/03_neural_recsys/Explicit_Feedback_Neural_Recommender_System.ipynb [hunk 1]
```diff
@@ -288,25 +288,30 @@
...
+    "        super().__init__()\n",
+    "        \n",
+    "        self.user_embedding = Embedding(output_dim=embedding_size, input_dim=max_user_id + 1,\n",
+    "                                        input_length=1, name='user_embedding')\n",
+    "        self.item_embedding = Embedding(output_dim=embedding_size, input_dim=max_item_id + 1,\n",
+    "                                        input_length=1, name='item_embedding')\n",
+    "        \n",
+    "        # The following two layers don't have parameters.\n",
+    "        self.flatten = Flatten()\n",
+    "        self.dot = Dot(axes=1)\n",
```
Comment (ogrisel) [inline labs/03_neural_recsys/Explicit_Feedback_Neural_Recommender_System.ipynb]: Is it really useful to store layers without parameters as attribute on the model? Would instantiating those classes in the call function add overhead? If so wouldn't it be possible to use functions instead? I am not familiar with TF2 keras good practices, especially with the Model class API.

File: labs/03_neural_recsys/Explicit_Feedback_Neural_Recommender_System.ipynb [hunk 2]
```diff
@@ -460,28 +465,39 @@
...
-    "item_vecs = Flatten()(item_embedding)\n",
-    "\n",
-    "input_vecs = Concatenate()([user_vecs, item_vecs])\n",
-    "input_vecs = Dropout(0.99)(input_vecs)\n",
-    "\n",
-    "x = Dense(64, activation='relu')(input_vecs)\n",
-    "y = Dense(2, activation='tanh')(x)\n",
-    "\n",
-    "model = Model(inputs=[user_id_input, item_id_input], outputs=y)\n",
+    "class DeepRecoModel(Model):\n",
```
Comment (ogrisel) [inline labs/03_neural_recsys/Explicit_Feedback_Neural_Recommender_System.ipynb]: I would rather call it `DeepRegressionModel` for consistency as even the implicit feedback models as recommenders.

Alternatively we could have:

- `SimpleExplicitFeedbackModel`.
- `DeepExplicitFeedbackModel`,
- `SimpleImplicitFeedbackModel`,
- `DeepImplicitFeedbackModel`.

as you prefer. I am also fine with `RegressionModel` / `DeepRegressionModel` for explicit feedback and then `TripletModel` / `DeepTripletModel` for the implicit feedback.

File: labs/03_neural_recsys/solutions/deep_explicit_feedback_recsys.py [hunk 3]
```diff
@@ -1,32 +1,42 @@
 # For each sample we input the integer identifiers
 # of a single user and a single item
-user_id_input = Input(shape=[1], name='user')
-item_id_input = Input(shape=[1], name='item')
+class DeepRecoModel(Model):
```
Comment (ogrisel) [inline labs/03_neural_recsys/solutions/deep_explicit_feedback_recsys.py]: This would have to be updated accordingly.

File: labs/03_neural_recsys/Explicit_Feedback_Neural_Recommender_System.ipynb (hunk 1 above)
Comment (arthurdouillard) [reply labs/03_neural_recsys/Explicit_Feedback_Neural_Recommender_System.ipynb]: Assuming we are in eager mode (default), there would be an overhead. And unfortunatly Tensorflow doesn't provide a "functional api" (like PyTorch) of this layer.

It seems that the Google-recommended way is instantiating in the __ini__ (or build) any kind of layers, even those without parameters.

**Comment by arthurdouillard** [comment]
I've updated the model name as suggested.

**Comment by ogrisel** [comment]
Thanks very much @arthurdouillard and sorry for the slow reaction time...
